README
NAME
    Term::Shell::Enhanced - More functionality for Term::Shell

VERSION
    version 1.100820

SYNOPSIS
      use Term::Shell::Enhanced;
      my $shell = Term::Shell::Enhanced->new;
      $shell->print_greeting;
      $shell->cmdloop;

DESCRIPTION
    This class subclasses Term::Shell and adds some functionality.

METHODS
  DEFAULTS
    FIXME

  PROMPT_VARS
    FIXME

  catch_run
    FIXME

  cmd
    FIXME

  expand
    FIXME

  fini
    FIXME

  get_history_filename
    FIXME

  getopt
    FIXME

  help_alias
    FIXME

  help_apropos
    FIXME

  help_cd
    FIXME

  help_echo
    FIXME

  help_eval
    FIXME

  help_pwd
    FIXME

  help_quit
    FIXME

  help_set
    FIXME

  init
    FIXME

  postloop
    FIXME

  precmd
    FIXME

  print_greeting
    FIXME

  prompt_str
    FIXME

  run_
    FIXME

  run_alias
    FIXME

  run_apropos
    FIXME

  run_cd
    FIXME

  run_echo
    FIXME

  run_pwd
    FIXME

  run_quit
    FIXME

  run_set
    FIXME

  smry_alias
    FIXME

  smry_apropos
    FIXME

  smry_cd
    FIXME

  smry_echo
    FIXME

  smry_eval
    FIXME

  smry_pwd
    FIXME

  smry_quit
    FIXME

  smry_set
    FIXME

FEATURES
    The following features are added:

    "history"
        When the shell starts up, it tries to read the command history from
        the history file. Before quitting, it writes the command history to
        the history file - it does not append to it, it overwrites the file.

        The default history file name is the shell name - with non-word
        characters replaced by underscores -, followed by "_history", as a
        dotfile in $ENV{HOME}. For example, if you shell's name is "mysh",
        the default history file name will be "~/.mysh_history".

        You can override the history file name in the "DEFAULTS()", like
        this:

            use constant DEFAULTS => (
                history_filename => ...,
                ...
            );

    "alias replacement"
        See the "alias" command below.

    "prompt strings"
        When subclassing Term::Shell::Enhanced, you can define how you want
        your prompt to look like. Use "DEFAULTS()" to override this.

            use constant DEFAULTS => (
                prompt_spec => ...,
                ...
            );

        You can use the following prompt variables:

            h    the hostname
            n    the shell name
            '#'  the command number (increased after each command)
            \\   a literal backslash

        You can extend the list of available prompt variables by defining
        your own PROMPT_VARS() - they are cumulative over the class
        hierarchy.

            use constant PROMPT_VARS => (
                key => value,
                ...
            );

        Since more elaborate prompt variables will have some interaction
        with the shell object, you might need a more elaborate
        "PROMPT_VARS()" definition:

            sub PROMPT_VARS {
                my $self = shift;
                (
                    key => $self->some_method,
                    ...
                );
            }

        The prompt variables are interpolated anew for every prompt.

        The default prompt string is:

            ': \n:\#; ',

        so if your shell is called "mysh", the default prompt looks somewhat
        like this:

           : mysh:1;

COMMANDS
    The following commands are added:

    "eval"
        You can evaluate snippets of Perl code just by putting them on a
        line beginning with "!":

          psh:~> ! print "$_\n" for keys %ENV

    "set [name[=value] ... ]"
        "set" lets you manipulate environment variables. You can view
        environment variables using "set". To view specific variables, use
        "set name". To set environment variables, use "set foo=bar".

    "cd [dir]"
          cd foo/bar/baz

        Change the current directory to the given directory. If no directory
        is given, the current value of $HOME is used.

    "pwd"
        Prints the current working directory.

    "alias [ name[=value] ... ]"
        "alias" with no arguments prints the list of aliases in the form
        "NAME=VALUE" on standard output. An alias is defined for each "NAME"
        whose "VALUE" is given.

        When you enter any command, it is checked against aliases and
        replaced if there is an alias defined for it. Only the command name
        - that is, the first word of the input line - undergoes alias
        replacement.

    "echo [arg ...]"
        Output the args.

    "quit"
        Exits the program.

    "apropos <word">
        Like the "help" command, but limits the information to commands that
        contain the given word in the command name or the summary.

INSTALLATION
    See perlmodinstall for information and options on installing Perl
    modules.

BUGS AND LIMITATIONS
    No bugs have been reported.

    Please report any bugs or feature requests through the web interface at
    <http://rt.cpan.org/Public/Dist/Display.html?Name=Term-Shell-Enhanced>.

AVAILABILITY
    The latest version of this module is available from the Comprehensive
    Perl Archive Network (CPAN). Visit <http://www.perl.com/CPAN/> to find a
    CPAN site near you, or see
    <http://search.cpan.org/dist/Term-Shell-Enhanced/>.

    The development version lives at
    <http://github.com/hanekomu/Term-Shell-Enhanced/>. Instead of sending
    patches, please fork this project using the standard git and github
    infrastructure.

AUTHOR
      Marcel Gruenauer <marcel@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2005 by Marcel Gruenauer.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl extension Term-Shell-Enhanced

1.100820  2010-03-23 17:45:04 Europe/Vienna
    - converted the distribution to Dist::Zilla-style

0.05  Thu 2010.02.11 15:00:29 CET (Marcel Gruenauer <marcel@cpan.org>)
     - Changed dist style and Changes back to standard. People didn't like it -
       the nail that sticks out gets hammered down.
     - Added standard test files; this will also help with CPANTS.

0.04  Thu, 29 May 2008 12:08:22 -0000 (Marcel Gruenauer <marcel@cpan.org>)
     - Converted Changes file to YAML style
     - .shipit: added Twitter step
     - Makefile.PL: added auto_install() and process_templates()
     - lib/*: converted to template
     - updated MANIFEST
     - updated t/perlcriticrc
     - fixed pod typo ('unterscores')
     - moved generated documentation of interited methods to after
       documentation of this class' methods so the manpage is more readable
     - .shipit: fixed svk.tagpattern
     - tags: NEWFEATURE

0.03  Thu, 18 Oct 2007 10:09:39 +0200 (Marcel Gruenauer <marcel@cpan.org>)
     - simplified accessor declaration

0.02  Thu, 18 Oct 2007 10:09:39 +0200 (Marcel Gruenauer <marcel@cpan.org>)
     - fixed version requirement of Test::More

0.01  Wed, 03 Oct 2007 23:15:39 +0200 (Marcel Gruenauer <marcel@cpan.org>)
     - original version


inc/Test/Compile.pm
#line 1
package Test::Compile;

use warnings;
use strict;
use Test::Builder;
use File::Spec;
use UNIVERSAL::require;


our $VERSION = '0.03';


my $Test = Test::Builder->new;


sub import {
    my $self = shift;
    my $caller = caller;

   for my $func ( qw( pm_file_ok all_pm_files all_pm_files_ok ) ) {
        no strict 'refs';
        *{$caller."::".$func} = \&$func;
    }

    $Test->exported_to($caller);
    $Test->plan(@_);
}


sub pm_file_ok {
    my $file = shift;
    my $name = @_ ? shift : "Compile test for $file";

    if (!-f $file) {
        $Test->ok(0, $name);
        $Test->diag("$file does not exist");
        return;
    }

    my $module = $file;
    $module =~ s!^(blib/)?lib/!!;
    $module =~ s!/!::!g;
    $module =~ s/\.pm$//;

    my $ok = 1;
    $module->use;
    $ok = 0 if $@;

    my $diag = '';
    unless ($ok) {
        $diag = "couldn't use $module ($file): $@";
    }

    $Test->ok($ok, $name);
    $Test->diag($diag) unless $ok;
    $ok;
}


sub all_pm_files_ok {
    my @files = @_ ? @_ : all_pm_files();

    $Test->plan(tests => scalar @files);

    my $ok = 1;
    for (@files) {
        pm_file_ok($_) or undef $ok;
    }
    $ok;
}


sub all_pm_files {
    my @queue = @_ ? @_ : _starting_points();
    my @pm;

    while ( @queue ) {
        my $file = shift @queue;
        if ( -d $file ) {
            local *DH;
            opendir DH, $file or next;
            my @newfiles = readdir DH;
            closedir DH;

            @newfiles = File::Spec->no_upwards(@newfiles);
            @newfiles = grep { $_ ne "CVS" && $_ ne ".svn" } @newfiles;

            for my $newfile (@newfiles) {
                my $filename = File::Spec->catfile($file, $newfile);
                if (-f $filename) {
                    push @queue, $filename;
                } else {
                    push @queue, File::Spec->catdir($file, $newfile);
                }
            }
        }
        if (-f $file) {
            push @pm, $file if $file =~ /\.pm$/;
        }
    }
    return @pm;
}


sub _starting_points {
    return 'blib' if -e 'blib';
    return 'lib';
}


1;

__END__

#line 261



lib/Term/Shell/Enhanced.pm
use 5.008;
use strict;
use warnings;

package Term::Shell::Enhanced;
our $VERSION = '1.100820';

# ABSTRACT: More functionality for Term::Shell
use Sys::Hostname;
use Getopt::Long;
use Cwd;
use parent qw(
  Data::Inherited
  Term::Shell
  Class::Accessor::Complex
);
__PACKAGE__->mk_hash_accessors(qw(opt))->mk_accessors(
    qw(
      num hostname log name longname prompt_spec history_filename
      )
);

# These aren't the constructor()'s DEFAULTS()!  Because new() comes from
# Term::Shell, we don't have the convenience of the the 'constructor'
# MethodMaker-generated constructor. Therefore, Term::Shell::Enhanced defines
# its own mechanism.
sub DEFAULTS {
    my $self = shift;
    (   num         => 0,
        name        => 'mysh',
        longname    => 'My Custom Shell',
        prompt_spec => ': \n:\#; ',
        hostname    => ((split(/\./, hostname))[0]),
    );
}

sub get_history_filename {
    my $self     = shift;
    my $filename = $self->history_filename;
    return $filename if defined $filename;

    # Per default, the history file name is derived from the shell name, with
    # non-word characters suitably changed to make a sane filename.
    (my $name = $self->name) =~ s/\W/_/g;
    "$ENV{HOME}/.$name\_history";
}

sub init {
    my $self = shift;
    $self->SUPER::init(@_);
    my %args = @{ $self->{API}{args} };
    $self->log($args{log}) unless defined $self->log;
    $self->opt($args{opt}) unless defined $self->opt;
    my %defaults = $self->every_hash('DEFAULTS');
    while (my ($key, $value) = each %defaults) {
        $self->$key($value) unless defined $self->$key;
    }

    # Only now can we try to read the history file, because the
    # 'history_filename' might have been defined in the DEFAULTS().
    if ($self->{term}->Features->{setHistory}) {
        my $filename = $self->get_history_filename;
        if (-r $filename) {
            open(my $fh, '<', $filename)
              or die "can't open history file $filename: $!\n";
            chomp(my @history = <$fh>);
            $self->{term}->SetHistory(@history);
            close $fh or die "can't close history file $filename: $!\n";
        }
    }
}

sub print_greeting {
    my $self = shift;
    printf <<EOINTRO, $self->name, $self->longname, our $VERSION;

%s -- %s (v%s)

Type 'help' for help, 'help <command>' for more detailed help on a command.

EOINTRO
}

sub precmd {
    my $self = shift;
    my $hnd  = shift;
    my $cmd  = shift;
    my $args = shift;
    @$args = $self->expand(@$args);
}

sub expand {
    my $self = shift;
    for (@_) {

        # it's easier to do this here instead of in cmd() because the input
        # will already have been split into words, so we can use '^' in
        # regexes to do what we mean.
        s[^~][$ENV{HOME}];
        s[\$([_A-Za-z0-9]+)][$ENV{$1} || '']eg;
    }
    @_;
}

sub cmd {
    my $self = shift;
    my $line = shift;
    if ($line =~ /^(\w+)/) {
        my $word = $1;
        if (exists $self->{SHELL}{alias}{$word}) {
            $line =~ s/^$word/ $self->{SHELL}{alias}{$word} /g;
        }
    }
    $self->SUPER::cmd($line);
}

sub PROMPT_VARS {
    my $self = shift;
    (   h    => $self->hostname,
        n    => $self->name,
        '#'  => $self->num,
        '\\' => '\\',
    );
}

# Can't use every_hash, because that caches and we might need dynamic values,
# such as the prompt number ($self->num)
sub prompt_str {
    my $self = shift;
    $self->num($self->num + 1);
    my %prompt_vars = $self->every_hash('PROMPT_VARS', 1);    # no caching
    (my $prompt = $self->prompt_spec) =~ s/\\(.)/$prompt_vars{$1} || ''/ge;
    $prompt;
}

sub getopt {
    my ($self, $cmd, $getopt_spec, @args) = @_;
    local @ARGV = @args;
    my %opt;
    my @getopt_spec =
      ref $getopt_spec eq 'ARRAY' ? @$getopt_spec : $getopt_spec;
    GetOptions(\%opt, @getopt_spec) or return $self->run_help($cmd);
    wantarray ? %opt : \%opt;
}

# The empty command; this sub needs to be there or the shell would exit
sub run_ {
    my $self = shift;

    # don't let the empty command count
    $self->num($self->num - 1);
}
sub fini { }

sub postloop {
    my $self = shift;
    print "\n";
    if ($self->{term}->Features->{getHistory}) {
        my $filename = $self->get_history_filename;
        open(my $fh, '>', $filename)
          or die "can't open history file $filename for writing: $!\n";
        print $fh "$_\n" for grep { length } $self->{term}->GetHistory;
        close $fh or die "can't close history file $filename: $!\n";
    }
}

# ========================================================================
# External commands
# ========================================================================
sub smry_eval { "how to evaluate Perl code" }

sub help_eval {
    <<'END' }
You can evaluate snippets of Perl code just by putting them on a line
beginning with !:

  psh:~> ! print "$_\n" for keys %ENV

END
{
    my $eval_num = "000001";

    sub catch_run {
        my ($o, $command, @args) = @_;

        # Evaluate perl code if it's a ! line.
        if ($command =~ s/^!//) {
            (my $code = $o->line) =~ s/^!//;
            my $really_long_string = <<END;
package Term::Shell::Enhanced::namespace_$eval_num;
{
    no strict;
    eval "no warnings";
    local \$^W = 0;
    $code;
}
END
            {
                local *_;
                my ($eval_num, $o, $command, @args, $code);
                eval $really_long_string;
            }
            print "$@\n" if $@;
            $eval_num++;
        } elsif ($command =~ s/^@//) {

            # Real external commands.
            system($command, @args);
        } elsif ($command =~ s/^://) {

            # The noop; ignore it
        } else {
            print "unknown command\n";
        }
    }
}

# ========================================================================
# set
# ========================================================================
sub smry_set { 'set environment variables' }

sub help_set {
    <<'END' }
set: set [ name[=value] ... ]
    set lets you manipulate environment variables. You can view environment
    variables using 'set'. To view specific variables, use 'set name'. To set
    environment variables, use 'set foo=bar'.

END

sub run_set {
    my ($o, @args) = @_;
    if (@args) {
        for my $arg (@args) {
            my ($key, $val) = split /=/, $arg;
            if (defined $val) {
                $ENV{$key} = $val;
            } else {
                $val = $ENV{$key} || '';
                print "$key=$val\n";
            }
        }
    } else {
        my ($key, $val);
        while (($key, $val) = each %ENV) {
            print "$key=$val\n";
        }
    }
}

# ========================================================================
# cd
# ========================================================================
sub smry_cd { 'change working directory' }

sub help_cd {
    <<'END' }
cd: cd [dir]
    Change the current directory to the given directory. If no directory is
    given, the current value of $HOME is used.

END

sub run_cd {
    my ($o, $dir) = @_;
    $dir = $ENV{HOME} unless defined $dir;
    chdir $dir or do {
        print "$0: $dir: $!\n";
        return;
    };
    $ENV{PWD} = $dir;
}

# ========================================================================
# pwd
# ========================================================================
sub smry_pwd { 'print working directory' }

sub help_pwd {
    <<'END' }
pwd: cwd
    Prints the current working directory.

END

sub run_pwd {
    my $o = shift;
    print getcwd;
}

# ========================================================================
# alias
# ========================================================================
sub smry_alias { 'view or set command aliases' }

sub help_alias {
    <<'END' }
alias: [ name[=value] ... ]
    'alias' with no arguments prints the list of aliases in the form
    NAME=VALUE on standard output. An alias is defined for each NAME whose
    VALUE is given.

END

sub run_alias {
    my $o = shift;
    if (@_) {
        for my $a (@_) {
            my ($key, $val) = split /=/, $a;
            if (defined $val) {
                $o->{SHELL}{alias}{$key} = $val;
            } else {
                $val = $o->{SHELL}{alias}{$key};
                print "alias $key=$val\n" if defined $val;
                print "alias: '$key' not found\n" if not defined $val;
            }
        }
    } else {
        my ($key, $val);
        my %alias = %{ $o->{SHELL}{alias} || {} };
        for my $alias (sort keys %alias) {
            printf "alias %s=%s\n", $alias, $alias{$alias};
        }
    }
}

# ========================================================================
# echo
# ========================================================================
sub smry_echo { 'output the args' }

sub help_echo {
    <<END }
echo [arg ...]
  Output the args.

END

sub run_echo {
    my ($self, @args) = @_;
    my @exp = $self->expand(@args);
    defined $_ or $_ = '' for @exp;
    print "@exp\n" if @exp;
}

# ========================================================================
# quit
# ========================================================================
sub smry_quit { 'exits the program' }

sub help_quit {
    <<END }
quit
  Exits the program.

END

sub run_quit {
    my $self = shift;
    $self->run_exit;
}

# ========================================================================
# apropos
# ========================================================================
sub smry_apropos { 'like "help", but limited to a topic' }

sub help_apropos {
    <<END }
apropos <word>
  Like the "help" command, but limits the information to commands that contain
  the given word in the command name or the summary.

END

# The implementation is taken directly from the run_help() method.
sub run_apropos {
    my $self = shift;
    my $word = shift;
    $word = '' unless defined $word;
    print "Type 'help command' for more detailed help on a command.\n";
    my (%cmds, %docs);
    my %done;
    my %handlers;
    for my $h (keys %{ $self->{handlers} }) {
        next unless length($h);
        next
          unless grep { defined $self->{handlers}{$h}{$_} } qw(run smry help);
        my $dest = exists $self->{handlers}{$h}{run} ? \%cmds : \%docs;
        my $smry =
          exists $self->{handlers}{$h}{smry}
          ? $self->summary($h)
          : "undocumented";
        my $help =
          exists $self->{handlers}{$h}{help}
          ? (
            exists $self->{handlers}{$h}{smry}
            ? ""
            : " - but help available"
          )
          : " - no help available";
        $dest->{"    $h"} = "$smry$help";
    }
    my (%apropos_cmds, %apropos_docs);

    # retain only matching commands and docs descriptions
    for my $cmd (keys %cmds) {
        next if index("$cmd$cmds{$cmd}", $word) == -1;
        $apropos_cmds{$cmd} = $cmds{$cmd};
    }
    for my $doc (keys %docs) {
        next if index("$doc$docs{$doc}", $word) == -1;
        $apropos_docs{$doc} = $docs{$doc};
    }
    print "  Commands:\n" if %apropos_cmds;
    $self->print_pairs(
        [ sort keys %apropos_cmds ],
        [ map { $apropos_cmds{$_} } sort keys %apropos_cmds ],
        ' - ', 1
    );
    print "  Extra Help Topics: (not commands)\n" if %apropos_docs;
    $self->print_pairs(
        [ sort keys %apropos_docs ],
        [ map { $apropos_docs{$_} } sort keys %apropos_docs ],
        ' - ', 1
    );
}
1;


__END__
=pod

=head1 NAME

Term::Shell::Enhanced - More functionality for Term::Shell

=head1 VERSION

version 1.100820

=for stopwords cmd fini getopt postloop precmd

=head1 SYNOPSIS

  use Term::Shell::Enhanced;
  my $shell = Term::Shell::Enhanced->new;
  $shell->print_greeting;
  $shell->cmdloop;

=head1 DESCRIPTION

This class subclasses L<Term::Shell> and adds some functionality.

=head1 METHODS

=head2 DEFAULTS

FIXME

=head2 PROMPT_VARS

FIXME

=head2 catch_run

FIXME

=head2 cmd

FIXME

=head2 expand

FIXME

=head2 fini

FIXME

=head2 get_history_filename

FIXME

=head2 getopt

FIXME

=head2 help_alias

FIXME

=head2 help_apropos

FIXME

=head2 help_cd

FIXME

=head2 help_echo

FIXME

=head2 help_eval

FIXME

=head2 help_pwd

FIXME

=head2 help_quit

FIXME

=head2 help_set

FIXME

=head2 init

FIXME

=head2 postloop

FIXME

=head2 precmd

FIXME

=head2 print_greeting

FIXME

=head2 prompt_str

FIXME

=head2 run_

FIXME

=head2 run_alias

FIXME

=head2 run_apropos

FIXME

=head2 run_cd

FIXME

=head2 run_echo

FIXME

=head2 run_pwd

FIXME

=head2 run_quit

FIXME

=head2 run_set

FIXME

=head2 smry_alias

FIXME

=head2 smry_apropos

FIXME

=head2 smry_cd

FIXME

=head2 smry_echo

FIXME

=head2 smry_eval

FIXME

=head2 smry_pwd

FIXME

=head2 smry_quit

FIXME

=head2 smry_set

FIXME

=head1 FEATURES

The following features are added:

=over 4

=item C<history>

When the shell starts up, it tries to read the command history from the
history file. Before quitting, it writes the command history to the history
file - it does not append to it, it overwrites the file.

The default history file name is the shell name - with non-word characters
replaced by underscores -, followed by C<_history>, as a dotfile in
C<$ENV{HOME}>. For example, if you shell's name is C<mysh>, the default
history file name will be C<~/.mysh_history>.

You can override the history file name in the C<DEFAULTS()>, like this:

    use constant DEFAULTS => (
        history_filename => ...,
        ...
    );

=item C<alias replacement>

See the C<alias> command below.

=item C<prompt strings>

When subclassing Term::Shell::Enhanced, you can define how you want your prompt to look like. Use C<DEFAULTS()> to override this.

    use constant DEFAULTS => (
        prompt_spec => ...,
        ...
    );

You can use the following prompt variables:

    h    the hostname
    n    the shell name
    '#'  the command number (increased after each command)
    \\   a literal backslash

You can extend the list of available prompt variables by defining your own PROMPT_VARS() - they are cumulative over the class hierarchy.

    use constant PROMPT_VARS => (
        key => value,
        ...
    );

Since more elaborate prompt variables will have some interaction with the shell object, you might need a more elaborate C<PROMPT_VARS()> definition:

    sub PROMPT_VARS {
        my $self = shift;
        (
            key => $self->some_method,
            ...
        );
    }

The prompt variables are interpolated anew for every prompt.

The default prompt string is:

    ': \n:\#; ',

so if your shell is called C<mysh>, the default prompt looks somewhat like
this:

   : mysh:1; 

=back

=head1 COMMANDS

The following commands are added:

=over 4

=item C<eval>

You can evaluate snippets of Perl code just by putting them on a line
beginning with C<!>:

  psh:~> ! print "$_\n" for keys %ENV

=item C<set [name[=value] ... ]>

C<set> lets you manipulate environment variables. You can view environment
variables using C<set>. To view specific variables, use C<set name>. To set
environment variables, use C<set foo=bar>.

=item C<cd [dir]>

  cd foo/bar/baz

Change the current directory to the given directory. If no directory is given,
the current value of C<$HOME> is used.

=item C<pwd>

Prints the current working directory.

=item C<alias [ name[=value] ... ]>

C<alias> with no arguments prints the list of aliases in the form
C<NAME=VALUE> on standard output. An alias is defined for each C<NAME> whose
C<VALUE> is given.

When you enter any command, it is checked against aliases and replaced if
there is an alias defined for it. Only the command name - that is, the first
word of the input line - undergoes alias replacement.

=item C<echo [arg ...]>

Output the args.

=item C<quit>

Exits the program.

=item C<apropos <word>>

Like the C<help> command, but limits the information to commands that contain
the given word in the command name or the summary.

=back

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests through the web interface at
L<http://rt.cpan.org/Public/Dist/Display.html?Name=Term-Shell-Enhanced>.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see
L<http://search.cpan.org/dist/Term-Shell-Enhanced/>.

The development version lives at
L<http://github.com/hanekomu/Term-Shell-Enhanced/>.
Instead of sending patches, please fork this project using the standard git
and github infrastructure.

=head1 AUTHOR

  Marcel Gruenauer <marcel@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2005 by Marcel Gruenauer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



inc/Module/Install/StandardTests.pm
#line 1
package Module::Install::StandardTests;

use warnings;
use strict;
use File::Spec;

use base 'Module::Install::Base';


our $VERSION = '0.04';


sub use_standard_tests {
    my ($self, %specs) = @_;
    
    my %with = map { $_ => 1 } qw/compile pod pod_coverage perl_critic/;
    if (exists $specs{without}) {
        $specs{without} = [ $specs{without} ] unless ref $specs{without};
        delete $with{$_} for @{ $specs{without} };
    }

    $self->build_requires('Test::More');
    $self->build_requires('UNIVERSAL::require');

    # Unlike other tests, this is mandatory.
    $self->build_requires('Test::Compile');

    $self->write_standard_test_compile;    # no if; this is mandatory
    $self->write_standard_test_pod          if $with{pod};
    $self->write_standard_test_pod_coverage if $with{pod_coverage};
    $self->write_standard_test_perl_critic  if $with{perl_critic};
}


sub write_test_file {
    my ($self, $filename, $code) = @_;
    $filename = File::Spec->catfile('t', $filename);

    # Outdent the code somewhat. Remove first empty line, if any. Then
    # determine the indent of the first line. Throw that amount of indenting
    # away from any line. This allows you to indent the code so it's visually
    # clearer (see methods below) while creating output that's indented more
    # or less correctly. Smoke result HTML pages link to the .t files, so it
    # looks neater.

    $code =~ s/^ *\n//;
    (my $indent = $code) =~ s/^( *).*/$1/s;
    $code =~ s/^$indent//gm;

    print "Creating $filename\n";
    open(my $fh, ">$filename") or die "can't create $filename $!";

    my $perl = $^X;
    print $fh <<TEST;
#!$perl -w

use strict;
use warnings;

$code
TEST

    close $fh or die "can't close $filename $!\n";
    $self->realclean_files($filename);
}


sub write_standard_test_compile {
    my $self = shift;
    $self->write_test_file('000_standard__compile.t', q/
        BEGIN {
            use Test::More;
            eval "use Test::Compile";
            Test::More->builder->BAIL_OUT(
                "Test::Compile required for testing compilation") if $@;
            all_pm_files_ok();
        }
    /);
}


sub write_standard_test_pod {
    my $self = shift;
    $self->write_test_file('000_standard__pod.t', q/
        use Test::More;
        eval "use Test::Pod";
        plan skip_all => "Test::Pod required for testing POD" if $@;
        all_pod_files_ok();
    /);
}


sub write_standard_test_pod_coverage {
    my $self = shift;
    $self->write_test_file('000_standard__pod_coverage.t', q/
        use Test::More;
        eval "use Test::Pod::Coverage";
        plan skip_all =>
            "Test::Pod::Coverage required for testing POD coverage" if $@;
        all_pod_coverage_ok();
    /);
}


sub write_standard_test_perl_critic {
    my $self = shift;
    $self->write_test_file('000_standard__perl_critic.t', q/
        use FindBin '$Bin';
        use File::Spec;
        use UNIVERSAL::require;
        use Test::More;

        plan skip_all =>
            'Author test. Set $ENV{TEST_AUTHOR} to a true value to run.'
            unless $ENV{TEST_AUTHOR};

        my %opt;
        my $rc_file = File::Spec->catfile($Bin, 'perlcriticrc');
        $opt{'-profile'} = $rc_file if -r $rc_file;

        if (Perl::Critic->require('1.078') &&
            Test::Perl::Critic->require &&
            Test::Perl::Critic->import(%opt)) {

            all_critic_ok("lib");
        } else {
            plan skip_all => $@;
        }
    /);
}


1;

__END__

#line 249



Makefile.PL

use strict;
use warnings;

 BEGIN { require 5.008; } 

use ExtUtils::MakeMaker 6.11;



my %WriteMakefileArgs = (
                       'test' => {
                                   'TESTS' => 't/*.t'
                                 },
                       'NAME' => 'Term::Shell::Enhanced',
                       'DISTNAME' => 'Term-Shell-Enhanced',
                       'CONFIGURE_REQUIRES' => {
                                                 'ExtUtils::MakeMaker' => '6.11'
                                               },
                       'AUTHOR' => 'Marcel Gruenauer <marcel@cpan.org>',
                       'BUILD_REQUIRES' => {},
                       'ABSTRACT' => 'More functionality for Term::Shell',
                       'EXE_FILES' => [],
                       'VERSION' => '1.100820',
                       'PREREQ_PM' => {
                                        'Class::Accessor::Complex' => '0',
                                        'Getopt::Long' => '0',
                                        'Scalar::Util' => '0',
                                        'parent' => '0',
                                        'English' => '0',
                                        'Term::Shell' => '0',
                                        'Carp' => '0',
                                        'Test::More' => '0.88',
                                        'File::Find' => '0',
                                        'File::Temp' => '0',
                                        'Data::Inherited' => '0',
                                        'Cwd' => '0',
                                        'Sys::Hostname' => '0'
                                      },
                       'LICENSE' => 'perl'
                     );


delete $WriteMakefileArgs{LICENSE}
  unless eval { ExtUtils::MakeMaker->VERSION(6.31) };

WriteMakefile(%WriteMakefileArgs);



